Build Docker image and push to GHCR

br3ndonland opened the pull request:
## Description

Closes #58

Up to this point, the project has been set up as a [Docker action](https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action) referencing the Dockerfile.

https://github.com/pypa/gh-action-pypi-publish/blob/3fbcf7ccf443305955ce16db9de8401f7dc1c7dd/action.yml#L86-L88

The downside to using the Dockerfile for the action is that the Docker image must be built every time the action is used (#58).

This PR will set up the project to build the Docker image and push it to [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) (GHCR). This change will speed up user workflows every time the action is used because the workflows will simply pull the Docker image from GHCR instead of building again.

## Changes

### Build container image with GitHub Actions

This PR will build Docker images with the Docker CLI (`docker build`). Builds will include inline [cache metadata](https://docs.docker.com/reference/cli/docker/image/build/#cache-from) so layers can be reused by future builds.

This PR only proposes to build container images for `x86_64` (`linux/amd64`) because GitHub Actions Linux runners currently only support `x86_64` CPU architectures (actions/runner-images#5631), and this project only supports GitHub Actions Linux runners. The [README explains](https://github.com/pypa/gh-action-pypi-publish/blob/3fbcf7ccf443305955ce16db9de8401f7dc1c7dd/README.md?plain=1#L109-L130):

> Since this GitHub Action is docker-based, it can only be used from within GNU/Linux based jobs in GitHub Actions CI/CD workflows. This is by design and is unlikely to change due to a number of considerations we rely on.

### Push container image to GHCR

The workflow will log in to GHCR using the [built-in GitHub token](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) and push the Docker image. Workflow runs triggered by pull requests will build the Docker image and run the smoke tests but will not push the Docker image.

### Update action to pull container image from GHCR

Docker actions support pulling in pre-built Docker images by supplying a registry address to the `image:` key. The downside to this syntax is that **there's no way to specify the correct Docker tag** because [the GitHub Actions `image:` and `uses:` keys don't accept any context](https://github.com/orgs/community/discussions/27048#discussioncomment-3254423). For example, if a user's workflow has `uses: pypa/gh-action-pypi-publish@release/v1.8`, then the action should pull in a Docker image built from the `release/v1.8` ref, something like `ghcr.io/pypa/gh-action-pypi-publish:release-v1.8` (Docker tags can't have `/`).

```yml
# this works but the image tag can't be customized
runs:
  using: docker
  image: docker://ghcr.io/pypa/gh-action-pypi-publish:release-v1.8
```

```yml
# this doesn't work because `image:` doesn't support context
runs:
  using: docker
  image: docker://ghcr.io/pypa/gh-action-pypi-publish:${{ github.action_ref }}
```

The workaround is to switch the top-level `action.yml` to a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) that then calls the Docker action, substituting the correct image name and tag.

## Related

- [GitHub docs: Actions - Creating actions - Creating a composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action)
- [GitHub docs: Packages - Working with a GitHub Packages registry - Container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) (GHCR)
- https://github.com/actions/runner/issues/2473
- https://github.com/actions/toolkit/issues/1264
- https://github.com/orgs/community/discussions/26212
- https://github.com/orgs/community/discussions/27048
- https://github.com/orgs/community/discussions/41927
- https://github.com/orgs/community/discussions/63675

br3ndonland committed:
Build Docker image and push to GHCR

Up to this point, the project has been set up as a Docker action
referencing the Dockerfile. The downside to using the Dockerfile for the
action is that the Docker image must be built every time the action is
used.

This commit will set up the project to build the Docker image and push
it to GitHub Container Registry (GHCR). This change will speed up user
workflows every time the action is used because the workflows will
simply pull the Docker image from GHCR instead of building again.

Changes:

- Add required metadata to Dockerfile
- Build container image with GitHub Actions
- Push container image to GHCR

Docker actions support pulling in pre-built Docker images. The downside
is that there's no way to specify the correct Docker tag because the
GitHub Actions `image` and `uses:` keys don't accept any context.
For example, if a user's workflow has
`uses: pypa/gh-action-pypi-publish@release/v1.8`, then the action should
pull in a Docker image built from the `release/v1.8` branch, something
like `ghcr.io/pypa/gh-action-pypi-publish:release-v1.8` (Docker tags
can't have `/`). The workaround is to switch the top-level `action.yml`
to a composite action that then calls the Docker action, substituting
the correct image name and tag.

br3ndonland committed:
Fix pre-commit errors

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,30 @@
+    - name: Push Docker image to GHCR
+      if: github.event_name != 'pull_request'
+      run: |
+        echo ${{ secrets.GITHUB_TOKEN }} |
```
Doesn't this need a `\` at the end? Can this be split into two steps?

br3ndonland replied:
Separating the `docker login` and `docker push` steps is a good idea. I've pushed that change (bb0315f).

There's no `\` needed in the `docker login`. The `|` with next line indented implies that the statement continues on the next line. This syntax also helps stay under the 80 character `yamllint` line length limit.

webknjaz replied:
Thanks! I'll mark this thread as resolved.

webknjaz commented:
This looks.. intriguing! I don't remember if I ever considered combining composite+docker actions (I did play with having two composites in the same repo in the past, though).

I'll need to take some time to think about it and look through the patch more closely. Please, don't expect an immediate review, however it does look very promising at glance!

Originally I thought that I'd have a workflow where I trigger a release, that release adds a commit that hardcodes an update to `action.yml` with the "future" version tag, tags that commit and pushes it (post docker publish). It wouldn't be on the `main` branch, the tags would be on the orphaned leaves.

This looks like a better idea so far. Thanks again!

br3ndonland committed:
Separate `docker login` and `docker push`

https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1578694138

br3ndonland commented:
That sounds great. Take your time. Thanks for your consideration.

If you do decide to accept this change, I'm happy to help maintain the workflows in the future. Feel free to mention me @br3ndonland and I will help address any issues that come up.

webknjaz commented:
Thanks! I've hit "rebase" on the UI to get this on top of the recent changes/linting/lockfile bumps but haven't yet looked into it deeper.

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,33 @@
+          --tag $IMAGE
+    - name: Log in to GHCR
+      if: github.event_name != 'pull_request'
+      run: |
```
I usually prefer this syntax so that the whole thing is parsed as a single line:
```suggestion
      run: >-
```

br3ndonland replied:
Generally I don't use `run: >-` because it will fold the contents of the block into a single line. This can alter the behavior of the script within the block. On this line though, it makes no difference, so sure.

webknjaz commented on .github/actions/run-docker-container/action.yml:
```diff
@@ -0,0 +1,44 @@
+runs:
+  using: docker
+  image: "{{image}}"
+  args:
+  - ${{ inputs.user }}
+  - ${{ inputs.password }}
+  - ${{ inputs.repository-url }}
+  - ${{ inputs.packages-dir }}
+  - ${{ inputs.verify-metadata }}
+  - ${{ inputs.skip-existing }}
+  - ${{ inputs.verbose }}
+  - ${{ inputs.print-hash }}
```
These aren't actually used. Somebody contributed them, but I don't see them being needed. So perhaps we shouldn't keep them around anymore.

br3ndonland replied:
Aren't the inputs used by the entrypoint script?

https://github.com/pypa/gh-action-pypi-publish/blob/699cd6103f50bf5c3b2f070c70712d109c168e6c/twine-upload.sh#L37-L41

webknjaz replied:
@br3ndonland they are being read through the environment variables. This bit passes them as CLI arguments additionally but said interface isn't actually used anywhere. I think somebody contributed it and I accepted that PR, but there was never really any need for it. I believe it only brings confusion, and removing it would make that confusion go away.

It's probably a good idea to drop them from the current default branch separately, though.

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,33 @@
+
+on:   # yamllint disable-line rule:truthy
+  pull_request:
+  push:
+    branches: ["release/*", "unstable/*"]
+    tags: ["*"]
```
When we push the tag first, there's going to be some time when the new version is out already, but the container doesn't actually exist yet. Therefore, I prefer building my release automation processes around the `workflow_dispatch` trigger, so that I'm able to type in the target version in the UI, get the image published and only then push a tag, only when it's known that the image exists already. Let's try reworking this with similar philosophy in mind.

br3ndonland replied:
Thanks for clarifying. I've replaced the tags trigger with `workflow_dispatch` instead (049447a).

webknjaz commented on .github/workflows/self-smoke-test-action.yml:
```diff
@@ -3,8 +3,10 @@
 on:   # yamllint disable-line rule:truthy
-  push:
   pull_request:
+  workflow_run:
```
Two separate workflow runs is often hard to track. Instead, I adopted a practice of modularizing the workflow pieces as reusable workflows having the `reusable-` prefix in their names. This allows embedding everything in all the right places. Let's try this, WDYT?

br3ndonland replied:
I will return to this suggestion at a later time.

webknjaz replied:
I think we should be able to circle back to this already...

br3ndonland replied:
> Two separate workflow runs is often hard to track. Instead, I adopted a practice of modularizing the workflow pieces as reusable workflows having the `reusable-` prefix in their names. This allows embedding everything in all the right places. Let's try this, WDYT?

@webknjaz I removed `workflow_call` and switched to a reusable workflow instead.

Note that the required status checks for the repo will need to be updated.

webknjaz commented on .github/workflows/self-smoke-test-action.yml:
```diff
@@ -42,13 +47,11 @@ jobs:
-      uses: actions/checkout@v3
-      with:
-        path: test
+      uses: actions/checkout@v4
```
Can we not make version upgrades in the same PR? And I'd rather not change the testing logic/dirs unless it's required somehow.

br3ndonland replied:
Ok, I'll stick with `actions/checkout@v3` and keep the previous `test` path.

webknjaz commented on .pre-commit-config.yaml:
```diff
@@ -31,7 +31,7 @@ repos:
     - --builtin-schema
     - github-workflows-require-timeout
-    files: ^\.github/workflows/[^/]+$
+    files: ^\.github\/workflows/[^/]+$
```
Could you submit this separately?

br3ndonland replied:
I've removed this change. The existing regex (`^\.github/workflows/[^/]+$`) seems like it should work fine, but pre-commit was matching files in `.github/actions/` and raising errors. Dynamically generating the entire `action.yml` as you [suggested](https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1603950013) will avoid the problem.

webknjaz commented on Dockerfile:
```diff
@@ -3,6 +3,7 @@ FROM python:3.12-slim
 LABEL "maintainer" "Sviatoslav Sydorenko <wk+pypa@sydorenko.org.ua>"
 LABEL "repository" "https://github.com/pypa/gh-action-pypi-publish"
 LABEL "homepage" "https://github.com/pypa/gh-action-pypi-publish"
+LABEL "org.opencontainers.image.source" "https://github.com/pypa/gh-action-pypi-publish"
```
I'd accept this in a separate PR even before the rest is figured out/refactored.

br3ndonland replied:
Moved to separate PR (#241).

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,59 @@ branding:
+  steps:
+  - name: Reset path if needed
+    run: |
+      # Reset path if needed
```
Why would this be needed outside the container?

br3ndonland replied:
Because you've set up a test that modifies the `$PATH` (https://github.com/pypa/gh-action-pypi-publish/issues/112, 1350b8bd724c983bdf030b52cba047a6eea5dc6d).

https://github.com/pypa/gh-action-pypi-publish/blob/699cd6103f50bf5c3b2f070c70712d109c168e6c/.github/workflows/self-smoke-test-action.yml#L85-L90

webknjaz replied:
Ack. Marking as resolved.

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,59 @@ branding:
+      # Set repo and ref from which to run Docker container action
+      # to handle cases in which `github.action_` context is not set
+      # https://github.com/actions/runner/issues/2473
+      REF=${{ env.ACTION_REF || github.ref_name }}
+      REPO=${{ env.ACTION_REPO || github.repository }}
+      echo "ref=$REF" >>"$GITHUB_OUTPUT"
+      echo "repo=$REPO" >>"$GITHUB_OUTPUT"
+    shell: bash
+    env:
+      ACTION_REF: ${{ github.action_ref }}
+      ACTION_REPO: ${{ github.action_repository }}
```
Isn't it easier to do it like this?
```suggestion
      REF=${{ github.action_ref || github.ref_name }}
      REPO=${{ github.action_repository || github.repository }}
      echo "ref=${REF}" >>"${GITHUB_OUTPUT}"
      echo "repo=${REPO}" >>"${GITHUB_OUTPUT}"
    shell: bash
```

br3ndonland replied:
You would think so, but the syntax is needed to work around a limitation of composite actions. See:

https://github.com/actions/runner/issues/2473
https://github.com/github/docs/issues/25336#issuecomment-1736251764

webknjaz replied:
Oh, good to know! It's probably safer too.

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,59 @@ branding:
+    env:
+      ACTION_REF: ${{ github.action_ref }}
+      ACTION_REPO: ${{ github.action_repository }}
+  - name: Set Docker image name and tag
+    run: |
+      # Set Docker image name and tag
+      # if action run was triggered by a pull request to this repo,
+      # build image from Dockerfile because it has not been pushed to GHCR,
+      # else pull image from GHCR
+      if [[ $GITHUB_EVENT_NAME == "pull_request" ]] &&
+        [[ $GITHUB_REPOSITORY == "pypa/gh-action-pypi-publish" ]]; then
+        IMAGE="../../../Dockerfile"
+      else
+        REF=${{ steps.set-repo-and-ref.outputs.ref }}
+        REPO=${{ steps.set-repo-and-ref.outputs.repo }}
+        IMAGE="docker://ghcr.io/$REPO:${REF/'/'/'-'}"
+      fi
+      FILE=".github/actions/run-docker-container/action.yml"
+      sed -i -e "s|{{image}}|$IMAGE|g" "$FILE"
+    shell: bash
```
So I was thinking… Why do we need to check in `action.yml` to Git even if this modifies it anyway? Why not generate it all, then?

Also, I'd use Python instead of Bash here. Then, it'd be possible to have a dict with data and write it as YAML using `json.dump()` (because JSON is valid YAML, almost always).

br3ndonland replied:
That could work. I've added a Python script that generates the Docker action (9ae1850).

Regarding your [related comment](https://github.com/pypa/gh-action-pypi-publish/pull/230#issuecomment-2116081280):

> I think that generating the file is a good idea. It should be possible to write the file without bringing in the PyYAML dependency. But it's not that easy for reading it. Can we make use of yq somehow, and convert YAML to JSON this way, maybe?

I've started by using PyYAML but will think about a way to do this with `yq`.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on .github/actions/run-docker-container/action.yml:
With https://github.com/pypa/gh-action-pypi-publish/pull/230/files#r1603950013 in mind, it should be possible not to add this file/folder into Git but produce it dynamically on the fly. Inputs and most of the stuff could be sourced from the top-level `action.yml`, maybe.

br3ndonland replied:
See related comment: https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1616378154

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,59 @@ branding:
+      user: ${{ inputs.user }}
+      password: ${{ inputs.password }}
+      repository-url: ${{ inputs.repository-url || inputs.repository_url }}
+      packages-dir: ${{ inputs.packages-dir || inputs.packages_dir }}
```
This would break the deprecation messages, it seems. Have you checked?

br3ndonland replied:
I don't think so. The new inputs don't have defaults, so for example, `${{ inputs.repository-url || inputs.repository_url }}` will default to `inputs.repository_url` and `deprecationMessage` will be logged.

webknjaz replied:
Alright. Marking this as resolved.

webknjaz commented:
I'm done with the initial review. More is needed, but I'd rather accept what I can through separate PRs to make this one smaller. And the suggested refactoring could be done in parallel. I think that generating the file is a good idea. It should be possible to write the file without bringing in the PyYAML dependency. But it's not that easy for reading it. Can we make use of `yq` somehow, and convert YAML to JSON this way, maybe?

br3ndonland committed:
Generate Docker container action with Python

br3ndonland committed:
Use YAML block strip syntax (`>-`) where possible

br3ndonland committed:
Reset pre-commit `files:` regex

br3ndonland committed:
Add `workflow_dispatch` trigger for Docker builds

br3ndonland commented:
@webknjaz thank you for your detailed review. I've addressed most of your comments so far.

webknjaz commented:
Commits like [213c885ac41d769527ac150e2e633bb1ccd886d4](https://github.com/pypa/gh-action-pypi-publish/commit/213c885ac41d769527ac150e2e633bb1ccd886d4) aren't really necessary since Git would automatically absorb changes applied separately. FYI. In fact, this one may have a harmful effect — when the label change is merged into the default branch, and this one rebased on top, Git will keep the removal commit and attempt deleting the label :man_shrugging:

Anyway, we'll address this later on.

webknjaz commented on create-docker-action.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def set_image(event: str, ref: str, repo: str) -> str:
+    if event == 'pull_request' and 'gh-action-pypi-publish' in repo:
```
Why do you check for `'gh-action-pypi-publish' in repo`? How about forks that renamed it?

br3ndonland replied:
If the workflow was triggered by a PR to _this repo_ (pypa/gh-action-pypi-publish), then the action should build the Docker image. Else, the action should be using the GHCR Docker image.

This is important because workflows in other repos might run the marketplace action on PRs. For example, they might want to smoke test package uploads using DevPI like this:

```yml
name: Smoke test workflow in a different repo

on:
  pull_request:

env:
  devpi-password: abcd1234
  devpi-username: root
  devpi-port: 3141

jobs:
  smoke-test:
    runs-on: ubuntu-latest
    services:
      devpi:
        image: muccg/devpi
        env:
          DEVPI_PASSWORD: ${{ env.devpi-password }}
        ports:
          - 3141
    timeout-minutes: 2
    steps:
      - name: Check out the action locally
        uses: actions/checkout@v4
      - name: Install the packaging-related tools
        run: python3 -m pip install build twine
      - name: Create the stub package importable directory
        run: mkdir -pv src/test_package
      - name: Populate the stub package `__init__.py`
        run: echo '__version__ = "0.1"' > src/test_package/__init__.py
      - name: Populate the stub package `README.md`
        run: echo "# Test Package" > README.md
      - name: Populate the stub package `pyproject.toml`
        run: echo "$CONTENTS" > pyproject.toml
        env:
          CONTENTS: |
            [build-system]
            requires = [
              "setuptools == 65.6.3",
            ]
            build-backend = "setuptools.build_meta"

            [project]
            name = "test-package"
            version = "0.1"
            readme = "README.md"
      - name: Build the stub package sdist and wheel distributions
        run: python3 -m build
      - name: Register the stub package in devpi
        run: twine register dist/*.tar.gz
        env:
          TWINE_USERNAME: ${{ env.devpi-username }}
          TWINE_PASSWORD: ${{ env.devpi-password }}
          TWINE_REPOSITORY_URL: >-
            http://localhost:${{
              job.services.devpi.ports[env.devpi-port]
            }}/${{
              env.devpi-username
            }}/public/
      - name: Smoke-test the marketplace action
        uses: pypa/gh-action-pypi-publish@unstable/v1
        with:
          user: ${{ env.devpi-username }}
          password: ${{ env.devpi-password }}
          repository-url: http://devpi:${{ env.devpi-port }}/${{ env.devpi-username }}/public/
```

In that case, `github.event_name` would be `pull_request`, but the repo name would be different.

As far as handling PRs to/from renamed forks of _this repo_ (pypa/gh-action-pypi-publish), if you have some other way of specifying "this repo" that doesn't include `gh-action-pypi-publish`, I'm happy to consider it.

webknjaz replied:
> if you have some other way of specifying "this repo" that doesn't include `gh-action-pypi-publish`, I'm happy to consider it.

@br3ndonland yes, the usual check is for the repo owner, not the repo name. So `github.repository_owner == 'pypa'` would work. A more precise check could be `github.repository_owner_id == 178055147` (the ID is taken from https://api.github.com/repos/pypa/gh-action-pypi-publish).

*(https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)*

br3ndonland replied:
@webknjaz good suggestion. I've updated the Python script and `action.yml` to check the repo owner ID.

Just a small correction - the ID you quoted (`178055147`) is actually `github.repository_id`. `github.repository_owner_id` for PyPA is `647025`. This can also be seen at https://api.github.com/orgs/pypa.

webknjaz replied:
@br3ndonland oh, right. I think I meant to suggest that checking the `github.repository_id` would be even more precise since it'd remain immutable in the event of the repository transfer even, but lost track of it. Could you make use of it instead?

br3ndonland replied:
@webknjaz right, I was thinking the same thing. Updated to use `github.repository_id`.

webknjaz replied:
Brilliant, thanks! Marking it as resolved.

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,59 @@ branding:
+  - name: Create Docker container action
+    run: |
+      # Create Docker container action
+      python -m pip install -r requirements/github-actions.txt
```
This is not the right way of using constraint files. Here's what should be done instead:
```suggestion
      PIP_CONSTRAINT=requirements/github-actions.txt python -Im pip install --only-binary=:all: -r requirements/github-actions.in
```

br3ndonland replied:
If we go with the latest approach of dumping JSON to a `.yml` file, we won't need to deal with the PyYAML requirement.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on requirements/github-actions.txt:
```diff
@@ -0,0 +1,8 @@
+# This file is autogenerated by pip-compile with Python 3.12
+# by the following command:
+#
+#    pip-compile --allow-unsafe --config=../.pip-tools.toml --output-file=github-actions.txt --strip-extras github-actions.in
```
Compiling PyYAML on various VMs might be problematic, so I'd pass `--only-binary=:all:` via pip-opts or even put it into the config. This way, installing would be more predictable on different versions of Ubuntu VMs that GitHub has.

Another concern is that originally, I opted for a docker-based action so that the external runtime is not mutated (that's one of a few reasons). By installing things with `pip` we violate that promise. I was kinda hoping that it'd be possible to avoid using a third-party dependency at all.

br3ndonland replied:
If we go with the latest approach of dumping JSON to a `.yml` file, we won't need to deal with the PyYAML requirement.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,59 @@ branding:
+  - name: Create Docker container action
+    run: |
+      # Create Docker container action
+      python -m pip install -r requirements/github-actions.txt
```
Another thing to consider, provided that it's actually needed (see https://github.com/pypa/gh-action-pypi-publish/pull/230/files#r1619055808), is making a virtualenv managed by this action (in its checkout/tmp dir so that the end-user's projects aren't polluted).

br3ndonland replied:
If we go with the latest approach of dumping JSON to a `.yml` file, we won't need to deal with the PyYAML requirement.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on create-docker-action.py:
```diff
@@ -0,0 +1,79 @@
+action_path = pathlib.Path('.github/actions/run-docker-container/action.yml')
+action_path.parent.mkdir(parents=True, exist_ok=True)
+
+with action_path.open(mode='w', encoding='utf-8') as file:
+    yaml.dump(action, file, allow_unicode=True, sort_keys=False)
```
Could you check if using just
```suggestion
action_path.write_text(json.dumps(action), encoding='utf-8')
```
would work?

Most of the time, any valid JSON is *also* valid YAML so GHA should be able to read it...

webknjaz replied:
Alternatively, we could have a template for the YAML file that would be post-processed by [Template strings](https://docs.python.org/3/library/string.html#template-strings), [`str.format_map()`](https://docs.python.org/3/library/stdtypes.html#str.format_map) or even just [f-strings](https://docs.python.org/3/reference/lexical_analysis.html#f-strings).

Any of these would let us avoid having an external dependency.

br3ndonland replied:
I was able to get it to work by dumping JSON to a `.yml` file.

webknjaz replied:
Sweet! Marking this thread as resolved.

webknjaz commented:
@br3ndonland I'm sure my review is incomplete, but hopefully it gives you enough ideas to try out until the next time.

I think it'd be nice to get this in before #236 if at all possible.

webknjaz commented:
@br3ndonland could you also rebase this branch locally? The GH button doesn't work, meaning there's going to be some conflicts to resolve.

br3ndonland commented:
> @br3ndonland could you also rebase this branch locally? The GH button doesn't work, meaning there's going to be some conflicts to resolve.

No problem. I'm not sure what "the GH button" is (some GitHub Mobile thing?) but I've rebased the branch.

I will address the other comments soon. Looks like there's a small `git fetch` issue in the smoke test workflow so I will look into that as well.

br3ndonland committed:
Checkout `github.head_ref` and repo for PRs

https://github.com/actions/checkout/issues/27#issuecomment-535897113
https://github.com/actions/checkout/issues/1108

br3ndonland committed:
Dump action as JSON

br3ndonland commented:
I was able to get it to work by dumping JSON to a `.yml` file. The file must have a `.yml` or `.yaml` extension (GitHub Actions will raise "Error: Can't find 'action.yml', 'action.yaml' or 'Dockerfile'" on `.json` files), but the file contents can be JSON.

webknjaz commented:
> I was able to get it to work by dumping JSON to a `.yml` file. The file must have a `.yml` or `.yaml` extension (GitHub Actions will raise "Error: Can't find 'action.yml', 'action.yaml' or 'Dockerfile'" on `.json` files), but the file contents can be JSON.

Great! That's what I expected since that JSON is still valid YAML and a YAML parser can read it.

webknjaz commented:
> > 
> 
> No problem. I'm not sure what "the GH button" is (some GitHub Mobile thing?) but I've rebased the branch.

Oh, it's usually present on the web UI. Might be missing from mobile. Sometimes its presence is conditional. It shows up somewhere right to the “Merge” button in PRs where you have the write privilege.

br3ndonland committed:
Check repo owner ID instead of repo name

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,64 @@ branding:
-  - ${{ inputs.verbose }}
-  - ${{ inputs.print-hash }}
+  using: composite
+  steps:
```
I've been meaning to suggest that we fail early here in case the runner runtime is unsupported. Originally, GitHub Actions would error out for docker actions on invocation. But since the wrapper is now composite, it'd be nice to do this manually so that it won't waste CPU on platforms where it won't function anyway.

```suggestion
  steps:
  - name: Fail-fast in unsupported environments
    if: runner.os != 'Linux'
    run: |
      >&2 echo This action is only able to run under GNU/Linux environments
      exit 1
    shell: bash -eEuo pipefail {0}
```

webknjaz commented on create-docker-action.py:
```diff
@@ -0,0 +1,78 @@
+    'runs': {
+        'using': 'docker',
+        'image': image,
+        'args': [
+            '${{ inputs.user }}',
+            '${{ inputs.password }}',
+            '${{ inputs.repository-url }}',
+            '${{ inputs.packages-dir }}',
+            '${{ inputs.verify-metadata }}',
+            '${{ inputs.skip-existing }}',
+            '${{ inputs.verbose }}',
+            '${{ inputs.print-hash }}',
+        ],
```
I think nothing would change if we were to drop this. The positional arguments are not being used.
```suggestion
```

br3ndonland replied:
Please see my previous comment (https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1616374020). Aren't the inputs used by the entrypoint script?

https://github.com/pypa/gh-action-pypi-publish/blob/699cd6103f50bf5c3b2f070c70712d109c168e6c/twine-upload.sh#L37-L41

webknjaz replied:
@br3ndonland I actually explained in the following comment https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1632408156 that those are read from environment variables and not from the positional CLI arguments:
https://github.com/pypa/gh-action-pypi-publish/blob/699cd6103f50bf5c3b2f070c70712d109c168e6c/twine-upload.sh#L21-L34

webknjaz replied:
If you look into the log, you'll see these as both env vars and pos args: https://github.com/pypa/gh-action-pypi-publish/actions/runs/9440557111/job/26000047753?pr=230#step:11:433.

br3ndonland replied:
Ok right, thanks for explaining that. I accepted the suggestion.

webknjaz commented on create-docker-action.py:
```diff
@@ -0,0 +1,78 @@
+    DESCRIPTION: (
+        'Run Docker container to upload Python distribution packages to PyPI'
+    ),
+    'inputs': {
```
Since Python 3.11, there's `tomllib` in stdlib. Let's use it to read the contents of `action.yml` and paste the `inputs` parsed from it here.

br3ndonland replied:
> Since Python 3.11, there's `tomllib` in stdlib. Let's use it to read the contents of `action.yml` and paste the `inputs` parsed from it here.

I don't think that will work because the file is not formatted as TOML. Attempting to load the file raises `tomllib.TOMLDecodeError`.

```text
gh-action-pypi-publish on ghcr via 🐍 v3.12.3
❯ python create-docker-action.py
Traceback (most recent call last):
  File "/Users/brendon/dev/gh-action-pypi-publish/create-docker-action.py", line 25, in <module>
    composite_action_contents = tomllib.loads(composite_action_text)
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/tomllib/_parser.py", line 102, in loads
    pos = key_value_rule(src, pos, out, header, parse_float)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/tomllib/_parser.py", line 326, in key_value_rule
    pos, key, value = parse_key_value_pair(src, pos, parse_float)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.3/Frameworks/Python.framework/Versions/3.12/lib/python3.12/tomllib/_parser.py", line 366, in parse_key_value_pair
    raise suffixed_err(src, pos, "Expected '=' after a key in a key/value pair")
tomllib.TOMLDecodeError: Expected '=' after a key in a key/value pair (at line 1, column 4)
```

webknjaz replied:
Right. I should stop reviewing in the middle of the night :)
Marking as resolved.

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,64 @@ branding:
+      ref: ${{ steps.set-repo-and-ref.outputs.ref }}
+      repository: ${{ steps.set-repo-and-ref.outputs.repo }}
+  - name: Create Docker container action
+    run: |
+      # Create Docker container action
+      python create-docker-action.py
```
No need to copy-paste the step description as a code comment, it doesn't add any value and would diverge over time.
```suggestion
    run: python create-docker-action.py
```

br3ndonland replied:
I'm fine with removing the actions step comments, but I don't agree that "it doesn't add any value." The comments are helpful because [GitHub Actions doesn't log steps clearly for composite actions](https://github.com/orgs/community/discussions/21276). It groups logs from all the steps together and doesn't log step names. The logs become difficult to read. The comments help because they are like step names.

To test this, I published to the GitHub Marketplace from my fork. Here's how the logs look when running the Marketplace action.

<img src="https://github.com/pypa/gh-action-pypi-publish/assets/26674818/ed3aa508-441a-4636-857f-013f523434a4" alt="Screenshot of GitHub Actions logs from composite action" width="75%" />

webknjaz replied:
Oh, I never thought of it like that! In some other actions, I've been trying to generate the groups manually. Alright, don't change it for now. I'm marking this thread as resolved.

P.S. You *don't have to* publish to Marketplace to test the action. You can reference any repository+committish (branch/tag/commit) regardless of whether it's on Marketplace or not.

br3ndonland committed:
Check repo ID instead of repo owner ID

br3ndonland committed:
Fail-fast in unsupported environments

https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1632406604

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

br3ndonland committed:
Drop args from create-docker-action.py

Co-authored-by: Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

br3ndonland commented:
@webknjaz thanks again for your detailed review so far.

How are things looking to you? Any further suggestions you would like me to look at? You mentioned reusable workflows https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1603922187. Anything else?

If/when you merge this PR, I am willing to switch my projects over to the `unstable/v1` branch to verify the action works as expected.

woodruffw commented on action.yml:
```diff
@@ -84,14 +84,70 @@ branding:
+  - name: Create Docker container action
+    run: |
+      # Create Docker container action
+      python create-docker-action.py
```
Sorry if I'm missing something here: why is creating the container action done dynamically like this? Is there a reason it can't be a static file?

br3ndonland replied:
Yes, the reason is that otherwise there's no way to specify the correct Docker tag.

Docker actions support pulling in pre-built Docker images by supplying a registry address to the `image:` key. The downside to this syntax is that **there's no way to specify the correct Docker tag** because [the GitHub Actions `image:` and `uses:` keys don't accept any context](https://github.com/orgs/community/discussions/27048#discussioncomment-3254423). For example, if a user's workflow has `uses: pypa/gh-action-pypi-publish@release/v1.8`, then the action should pull in a Docker image built from the `release/v1.8` ref, something like `ghcr.io/pypa/gh-action-pypi-publish:release-v1.8` (Docker tags can't have `/`).

```yml
# this works but the image tag can't be customized
runs:
  using: docker
  image: docker://ghcr.io/pypa/gh-action-pypi-publish:release-v1.8
```

```yml
# this doesn't work because `image:` doesn't support context
runs:
  using: docker
  image: docker://ghcr.io/pypa/gh-action-pypi-publish:${{ github.action_ref }}
```

The workaround is to switch the top-level `action.yml` to a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) that then creates the Docker container action, substituting the correct image name and tag.

Originally this PR proposed to create the Docker container action as a YAML file with a single placeholder field, like `image: {{image}}`, then replace the `{{image}}` placeholder with the image tag using `sed`. The maintainer preferred to have the entire Docker container action generated with a Python script instead.

webknjaz replied:
@woodruffw I requested that. I don't want to have another file with hard-coded contents in the repo. There was a template-like thing before. The published container version is being generated dynamically.

webknjaz commented on action.yml:
```diff
@@ -84,14 +84,70 @@ branding:
+    if: runner.os != 'Linux'
+    run: |
+      >&2 echo This action is only able to run under GNU/Linux environments
+      exit 1
```
Could you add CI jobs testing that this fails under corresponding envs?

br3ndonland replied:
I added CI jobs that test failure on macOS and Windows.

webknjaz replied:
Marking this as resolved.

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,39 @@
+    branches: ["release/*", "unstable/*"]
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: Docker image tag
```
Can we incorporate the prior art from the discussions in #45 and accept the 3-segment version as an input, then deduce everything else from that + push the Git tag and advance the proper branches? Perhaps, this could also happen before this PR so it exists as a separate base.

br3ndonland replied:
> Can we incorporate the prior art from the discussions in #45 and accept the 3-segment version as an input, then deduce everything else from that + push the Git tag and advance the proper branches? Perhaps, this could also happen before this PR so it exists as a separate base.

I don't understand what you're asking here. You want separate inputs for major, minor, and patch versions? What if you're building the Docker image from the `release/v1` branch? What is the "3-segment version" in that case?

webknjaz replied:
@br3ndonland I was talking about the workflow inputs. Everything else should be computed from them.

The current workflow for me is as follows:
1) Tag `v1.10.1` on top of `unstable/v1`
2) Fast-forward `release/v1` to `v1.10.1`
3) Fast-forward `release/v1.10` to `v1.10.1` (or create it)
4) `git push --atomic` everything in one go
5) Create a GitHub Release from the UI

With this, my 3-segment “input” is `1.10.1` and `v1`+`v1.10` are being extracted from that. Do you think we could bake containers into this flow with not many changes? This will involve tagging the same image with multiple tags, as I understand. And the end-users referencing versions by both tags and branches should be able to retrieve whatever container there is, right?

br3ndonland replied:
@webknjaz thanks for explaining that. I think we're actually pretty close to the desired workflow. There's just a small update needed to the Docker build workflow so it pushes tags with the major and minor version numbers. I've pushed that change.

Your updated workflow would look like this:

1. Tag `v1.10.1` on top of `unstable/v1`
2. Fast-forward `release/v1` to `v1.10.1`
3. Fast-forward `release/v1.10` to `v1.10.1` (or create it)
4. `git push --atomic` everything in one go
5. Trigger the :building_construction: workflow from the Git tag. The workflow will push the Docker image with three Docker tags:
   1. `ghcr.io/pypa/gh-action-pypi-publish:v1.10.1`
   2. `ghcr.io/pypa/gh-action-pypi-publish:v1.10`
   3. `ghcr.io/pypa/gh-action-pypi-publish:v1`
6. Create a GitHub Release from the UI

webknjaz replied:
@br3ndonland so with #45, I wanted to move all these steps inside the automation. For publishing Python projects, I usually use a workflow with `workflow_dispatch` that allows me to type in the desired version to release and that workflow creates the git tag, the gh release, signs stuff and so on.

The idea is that tag+branches+gh-release represent the result of release automation being successful and aren't triggers or manual actions.

br3ndonland replied:
> @br3ndonland so with #45, I wanted to move all these steps inside the automation. For publishing Python projects, I usually use a workflow with `workflow_dispatch` that allows me to type in the desired version to release and that workflow creates the git tag, the gh release, signs stuff and so on.
> 
> The idea is that tag+branches+gh-release represent the result of release automation being successful and aren't triggers or manual actions.

So... you want the Docker build workflow to run automatically when a tag is created? Or when a GitHub Release is created?

Please be more specific here, and please limit the scope of your requests. The goal of this PR is "Build Docker image and push to GHCR." The goal is not to develop your entire release workflow.

PR #45 is currently a draft and I don't think I should have to make this PR dependent on some other draft PR.

webknjaz replied:
> So... you want the Docker build workflow to run automatically when a tag is created? Or when a GitHub Release is created?

No to both. In my release processes, tags + GH Releases are the results of publishing, not the triggers. So when I trigger `workflow_dispatch` (which is a button click on GH UI + the desired version entered into the input field), the workflows create what I want to publish, send that to the primary platform (PyPI in case of Python projects or a container registry in the context of this specific PR), and only after that the same workflow creates the Git tag and the GH Release.

> Please be more specific here, and please limit the scope of your requests. The goal of this PR is "Build Docker image and push to GHCR." The goal is not to develop your entire release workflow.
> 
> PR #45 is currently a draft and I don't think I should have to make this PR dependent on some other draft PR.

I only pointed to that PR as this one seems to be touching the same topic partially. I was thinking that since you're already making a workflow that publishes the container, that could piggyback on top of this effort. I wasn't thinking of it as a dependency but rather as a replacement, but wanted to point to it for the pre-existing context around what I wanted to implement for quite a while.

This PR is attempting to change my release workflow already, after all. So I wanted to make sure it doesn't go in the direction that would complicate things, given that I already envision the release process to develop in a certain way.

If it were possible to merge the PR and keep the workflow where the Git tag is pushed, and it works right away without a container published for a while, it'd be where we could scope it. However, AFAIU, this will no longer work after merging since when there's a tag and the end-users start using it, the action will attempt pulling the image that is not yet there. There's a race condition in this process, meaning that it is prone to human error and needs to be handled carefully. And that's why I was thinking that it'd be good to integrate all the other things in a way that would keep the process robust.

P.S. Would it be useful to still give the end-users some way to force building the container instead of pulling the cached one?

br3ndonland committed:
Verify fail-fast in unsupported environments

webknjaz commented:
@br3ndonland this also needs to be rebased now that the attestations PR is in.

br3ndonland committed:
Add Docker tags for major and minor versions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,53 @@
+        docker push $IMAGE
+        docker push $IMAGE_MAJOR
+        docker push $IMAGE_MAJOR_MINOR
+  smoke-test:
```
There should be a dependency on the previous job
```suggestion
  smoke-test:
    needs:
    - build-and-push
```

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,53 @@
+    inputs:
+      tag:
+        description: Docker image tag (optional, defaults to Git ref)
+        required: false
```
I think this should be required. I don't want the version to be guessed as there's no good heuristics until we have change notes that would indicate breaking changes or something.

br3ndonland replied:
The version isn't "guessed," it's in the name of the Git tag that's used to trigger the workflow.

But sure, if you want to type the version number again, we can make that field required.

Also regarding https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1759783982, I'm still not sure what you're looking for here.

webknjaz replied:
@br3ndonland with `workflow_dispatch`, the version isn't available since it's not triggered by tag manipulations and so it must be an input. I'm specifically avoiding tags being the triggers. So the process is inverted. But it looks like you're already doing that (I haven't yet checked the new commits).

br3ndonland committed:
Make `workflow_dispatch` Docker tag input required

https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1759496153

webknjaz commented on .github/workflows/build-and-push-docker-image.yml:
```diff
@@ -0,0 +1,55 @@
+        echo ${{ secrets.GITHUB_TOKEN }} |
+          docker login ghcr.io -u $GITHUB_ACTOR --password-stdin
+    - name: Push Docker image to GHCR
+      if: github.event_name != 'pull_request'
```
Is it possible to pass the image between containers? It would be a good idea to pass it to the tests job and then to another one that would actually publish it, as opposed to publishing and then testing whether it should've been published in the first place...

br3ndonland replied:
Do you mean, "pass the image between _workflow jobs_?" Not directly. A different workflow job is a different virtual machine. We could [upload and download artifacts](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/storing-and-sharing-data-from-a-workflow), but that seems like unnecessary complexity.

What about just running the smoke tests before the Docker builds?

webknjaz replied:
Yeah, I was imagining using the artifacts. In the past there was one project where I was publishing to GHCR from one job with some unique tag and then pulling from the following one. But using artifacts is better.

It's indeed more complex, but that's how I build testing these days. OTOH, since you asked to cut the scope, I'll consider this a future improvement idea for now.

webknjaz commented on action.yml:
```diff
@@ -91,15 +91,71 @@ branding:
+      # Set repo and ref from which to run Docker container action
+      # to handle cases in which `github.action_` context is not set
+      # https://github.com/actions/runner/issues/2473
+      REF=${{ env.ACTION_REF || env.PR_REF || github.ref_name }}
```
@br3ndonland when running in another repo, wouldn't `github.ref_name` point to that repo instead of ours? Is it safe to use this var as a fallback? Same question for `github.event.pull_request`. Is this to cover contributions to the action itself?

br3ndonland replied:
If another repo is using the marketplace action (`uses: pypa/gh-action-pypi-publish`), `env.ACTION_REF` will be set from [`github.action_ref`](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#github-context).

If the action is being run on a PR to this repo, `env.PR_REF` will be set from `github.event.pull_request.head.ref`.

If the action is being run from a push, release, or other event in this repo, `github.ref_name` will be used.

webknjaz replied:
Alright. Thanks for clarifying. So I'll need to be careful in the release process to make an image before tagging…

webknjaz commented:
@br3ndonland with https://github.com/pypa/gh-action-pypi-publish/pull/230/files#r1787026278 in mind, if I switch to the process where I trigger the GHA workflow, wait for the container to show up and then manually tag the same commit, would that cover everything necessary post this patch? If yes, I think it'd fine to stop making the changes and merge this. I'll still want to see the clarifications on that question about reference fallbacks, though. And the PR needs rebasing to resolve the conflicts.

br3ndonland committed:
Move smoke test to reusable workflow

br3ndonland committed:
Run smoke tests before Docker builds

https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1787027821

webknjaz commented:
@br3ndonland so I was almost ready to merge this and realized that I don't understand a few things. Could you help me get some clarity?

1) When `workflow_dispatch` is triggered, and I type in `1.12.0`, I understand that it'll publish `ghcr.io/pypa/gh-action-pypi-publish:v1.12.0`, `ghcr.io/pypa/gh-action-pypi-publish:v1.12` and `ghcr.io/pypa/gh-action-pypi-publish:v1`. But how about `ghcr.io/pypa/gh-action-pypi-publish:release-v1` and `ghcr.io/pypa/gh-action-pypi-publish:unstable-v1`? What's the process for those?
2) Should the case of `uses: pypa/gh-action-pypi-publish@unstable/v1` not hit the GHCR image and build the image like it does now? Does `set_image()` need to be modified to implement this?
3) Should the case of `uses: pypa/gh-action-pypi-publish@release/v1` hit the `:v1` image, or should the image generation workflow produce one more tag? It looks like these two bits aren't cohesive: reading the diff once again, it looks like if I enter `release/v1.11.0`, it'll produce `ghcr.io/pypa/gh-action-pypi-publish:release-v1`, `ghcr.io/pypa/gh-action-pypi-publish:release-v1.11` and `ghcr.io/pypa/gh-action-pypi-publish:release-v1.11.0` which would then not match references like `uses: pypa/gh-action-pypi-publish@v1.11.0`.

Am I missing anything?

br3ndonland commented:
> 1. When `workflow_dispatch` is triggered, and I type in `1.12.0`, I understand that it'll publish `ghcr.io/pypa/gh-action-pypi-publish:v1.12.0`, `ghcr.io/pypa/gh-action-pypi-publish:v1.12` and `ghcr.io/pypa/gh-action-pypi-publish:v1`.

At your request (https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1759496153), the `workflow_dispatch` menu has two fields:

1. Git ref (branch or tag)
2. Docker image tag

After navigating to the [action](https://github.com/pypa/gh-action-pypi-publish/actions/workflows/build-and-push-docker-image.yml), there will be a "Run workflow" button. To publish an image for a specific release, v1.13.0 for example:

1. Git ref - v1.13.0 (tag)
2. Docker image tag - v1.13.0

> But how about `ghcr.io/pypa/gh-action-pypi-publish:release-v1` and `ghcr.io/pypa/gh-action-pypi-publish:unstable-v1`? What's the process for those?

Currently the workflow will automatically build and push Docker images whenever new commits are pushed to `unstable/*` or `release/*` branches.

https://github.com/pypa/gh-action-pypi-publish/blob/36965cb24aed76043aeb8f19062b6a31fa56b990/.github/workflows/build-and-push-docker-image.yml#L5-L8

> 2. Should the case of `uses: pypa/gh-action-pypi-publish@unstable/v1` not hit the GHCR image and build the image like it does now? Does `set_image()` need to be modified to implement this?

Why wouldn't you want to push a Docker image for `unstable/v1`? I wouldn't want that as a user. It seems like it would just introduce more confusion and complication.

> 3. Should the case of `uses: pypa/gh-action-pypi-publish@release/v1` hit the `:v1` image, or should the image generation workflow produce one more tag? It looks like these two bits aren't cohesive: reading the diff once again, it looks like if I enter `release/v1.11.0`, it'll produce `ghcr.io/pypa/gh-action-pypi-publish:release-v1`, `ghcr.io/pypa/gh-action-pypi-publish:release-v1.11` and `ghcr.io/pypa/gh-action-pypi-publish:release-v1.11.0` which would then not match references like `uses: pypa/gh-action-pypi-publish@v1.11.0`.

The case of `uses: pypa/gh-action-pypi-publish@release/v1` will use the code on the `release/v1` branch and will pull the Docker image built from that same branch.

> if I enter `release/v1.11.0`

Your repo does not have any branches or tags named `release/v1.11.0`, so why would you do that?

If you're publishing a Docker image for a specific release, enter the name of the tag as I said above.

1. Git ref - v1.13.0 (tag)
2. Docker image tag - v1.13.0

`uses: pypa/gh-action-pypi-publish@v1.13.0` will then pull the Docker image built from the v1.13.0 Git tag.

webknjaz commented:
> > 1. When `workflow_dispatch` is triggered, and I type in `1.12.0`, I understand that it'll publish `ghcr.io/pypa/gh-action-pypi-publish:v1.12.0`, `ghcr.io/pypa/gh-action-pypi-publish:v1.12` and `ghcr.io/pypa/gh-action-pypi-publish:v1`.
> 
> At your request ([#230 (comment)](https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1759496153)), the `workflow_dispatch` menu has two fields:
> 
>     1. Git ref (branch or tag)
> 
>     2. Docker image tag
> 
> 
> After navigating to the [action](https://github.com/pypa/gh-action-pypi-publish/actions/workflows/build-and-push-docker-image.yml), there will be a "Run workflow" button. To publish an image for a specific release, v1.13.0 for example:
> 
>     1. Git ref - v1.13.0 (tag)
> 
>     2. Docker image tag - v1.13.0

Are you sure you're looking at the latest PR diff? Right now, when I look at https://github.com/pypa/gh-action-pypi-publish/pull/230/files#diff-63d86cca28be8586869affd2baf5d48489229f5d9b4684c11beb787eee6f614cR9-R14, I see a single input — `tag`. Is it possible that you might have a branch with something different?

Though, reading through the workflow, the docker image tag is computed from the tag input, which makes sense to me.

> > But how about `ghcr.io/pypa/gh-action-pypi-publish:release-v1` and `ghcr.io/pypa/gh-action-pypi-publish:unstable-v1`? What's the process for those?
> 
> Currently the workflow will automatically build and push Docker images whenever new commits are pushed to `unstable/*` or `release/*` branches.
> 
> https://github.com/pypa/gh-action-pypi-publish/blob/36965cb24aed76043aeb8f19062b6a31fa56b990/.github/workflows/build-and-push-docker-image.yml#L5-L8

Oh… That's the bit I've been overlooking.

> > 2. Should the case of `uses: pypa/gh-action-pypi-publish@unstable/v1` not hit the GHCR image and build the image like it does now? Does `set_image()` need to be modified to implement this?
> 
> Why wouldn't you want to push a Docker image for `unstable/v1`? I wouldn't want that as a user. It seems like it would just introduce more confusion and complication.

Oh, I was just writing down my confusion. I didn't look into the PR in a while, so I didn't remember all the context and got lost. It makes sense now.

> > 3. Should the case of `uses: pypa/gh-action-pypi-publish@release/v1` hit the `:v1` image, or should the image generation workflow produce one more tag? It looks like these two bits aren't cohesive: reading the diff once again, it looks like if I enter `release/v1.11.0`, it'll produce `ghcr.io/pypa/gh-action-pypi-publish:release-v1`, `ghcr.io/pypa/gh-action-pypi-publish:release-v1.11` and `ghcr.io/pypa/gh-action-pypi-publish:release-v1.11.0` which would then not match references like `uses: pypa/gh-action-pypi-publish@v1.11.0`.
> 
> The case of `uses: pypa/gh-action-pypi-publish@release/v1` will use the code on the `release/v1` branch and will pull the Docker image built from that same branch.
> 
> > if I enter `release/v1.11.0`
> 
> Your repo does not have any branches or tags named `release/v1.11.0`, so why would you do that?
> 
> If you're publishing a Docker image for a specific release, enter the name of the tag as I said above.
> 
>     1. Git ref - v1.13.0 (tag)
> 
>     2. Docker image tag - v1.13.0
> 
> 
> `uses: pypa/gh-action-pypi-publish@v1.13.0` will then pull the Docker image built from the v1.13.0 Git tag.

This was very helpful to refresh my understanding of the bits and pieces of the updates in the PR. Thank you!

I think I'm ready to merge it finally. I'll then try it out and let you know if something is still uncertain.

webknjaz commented:
I just pushed `release/v1` and realized that people will be hitting a non-existing image if they attempt a release during the next few minutes… Hopefully, not many will hit this. For any following pushes, they'll just be getting a slightly outdated image, which should be mostly fine except for when the older image is incompatible.

webknjaz commented:
@br3ndonland here are some regressions reported:
* action pinned to a commit hash: https://github.com/pypa/gh-action-pypi-publish/issues/290
* self-hosted runner: https://github.com/pypa/gh-action-pypi-publish/issues/289
* nested composite action: https://github.com/pypa/gh-action-pypi-publish/issues/291
* GitHub Enterprise: https://github.com/pypa/gh-action-pypi-publish/issues/292

Could you help me figure them out?

webknjaz commented on action.yml:
```diff
@@ -91,15 +91,70 @@ branding:
+      PR_REPO: ${{ github.event.pull_request.head.repo.full_name }}
+      PR_REPO_ID: ${{ github.event.pull_request.base.repo.id }}
+  - name: Check out action repo
+    uses: actions/checkout@v4
```
@br3ndonland could you remind me — is this necessary because it should be within the work dir?

br3ndonland replied:
> @br3ndonland could you remind me — is this necessary because it should be within the work dir?

Initially I thought it was necessary because, when the composite action runs, `github.workspace` is _the working directory on the user's runner_, not the pypa/gh-action-pypi-publish repo. But I think `github.action_path` can be used to access files from the action repo, so this second checkout may not have been necessary. Hopefully `github.action_path` is still accurate for nested composite actions (https://github.com/pypa/gh-action-pypi-publish/issues/299).

webknjaz commented:
v1.12.1 addresses immediate problems from the list above. Long-term, the release process will need to be improved and made more cohesive.

br3ndonland commented:
> @br3ndonland here are some regressions reported:
>
> - action pinned to a commit hash: [Unable to find image 'ghcr.io/pypa/gh-action-pypi-publish:61da13deb5f5124fb1536194f82ed3d9bbc7e8f3' #290](https://github.com/pypa/gh-action-pypi-publish/issues/290)
> - self-hosted runner: [After this update, my github actions are broken with below logs #289](https://github.com/pypa/gh-action-pypi-publish/issues/289)
> - nested composite action: [No Such File or Directory #291](https://github.com/pypa/gh-action-pypi-publish/issues/291)
> - GitHub Enterprise: [Unable to use v1.12.0 on Github Enterprise #292](https://github.com/pypa/gh-action-pypi-publish/issues/292)
>
> Could you help me figure them out?

@webknjaz thanks for handling the user feedback so far. Here are some updates to the list you provided.

- action pinned to a commit hash: https://github.com/pypa/gh-action-pypi-publish/issues/290
  - Comment: https://github.com/pypa/gh-action-pypi-publish/issues/290#issuecomment-2469335732
  - Fix: https://github.com/pypa/gh-action-pypi-publish/pull/301
- self-hosted runner: https://github.com/pypa/gh-action-pypi-publish/issues/289
  - Fix: https://github.com/pypa/gh-action-pypi-publish/pull/294
- nested composite action: https://github.com/pypa/gh-action-pypi-publish/issues/291
  - I agree with your comment (https://github.com/pypa/gh-action-pypi-publish/issues/299#issuecomment-2462644247) that https://github.com/pypa/gh-action-pypi-publish/pull/293 does not necessarily fix this problem.
  - Possible duplicate: https://github.com/pypa/gh-action-pypi-publish/issues/299
  - Possible duplicate: https://github.com/pypa/gh-action-pypi-publish/issues/300
  - Related: https://github.com/pypa/gh-action-pypi-publish/pull/230#discussion_r1616376877
  - Fix: https://github.com/pypa/gh-action-pypi-publish/pull/302
- GitHub Enterprise: https://github.com/pypa/gh-action-pypi-publish/issues/292
  - There isn't enough information in #292 to fix the issue, but it was possibly occurring because GitHub Enterprise modifies [`$GITHUB_SERVER_URL`](https://github.com/actions/checkout/blob/3b9b8c884f6b4bb4d5be2779c26374abadae0871/src/url-helper.ts#L23-L30). If so, the fix would have been a simple one-line change to the `actions/checkout` step in the composite action to set `github-server-url: https://github.com`.
    ```yml
    - name: Check out action repo
      uses: actions/checkout@v4
      with:
        github-server-url: https://github.com
        path: action-repo
        ref: ${{ steps.set-repo-and-ref.outputs.ref }}
        repository: ${{ steps.set-repo-and-ref.outputs.repo }}
    ```
  - Fix: https://github.com/pypa/gh-action-pypi-publish/pull/293 (the second checkout step may not have been necessary because we can use `github.action_path`)
